src/libpriv/rpmostree-core.c

```diff
@@ -2854,8 +2870,37 @@ delete_package_from_root (RpmOstreeContext *self,
       if (!g_str_has_prefix (fn, "usr/"))
         continue;
 
-      if (!glnx_shutil_rm_rf_at (rootfs_dfd, fn, cancellable, error))
-        return FALSE;
+      /* Delete files first, we'll handle directories next */
+      if (S_ISDIR (mode))
+        g_sequence_insert_sorted (dirs_to_remove, g_strdup (fn), compare_strlen, NULL);
+      else
+        {
+          if (unlinkat (rootfs_dfd, fn, 0) < 0)
+            {
+              if (errno != ENOENT)
+                return glnx_throw_errno_prefix (error, "unlinkat(%s)", fn);
+            }
+        }
+    }
+
+  /* We want to perform a bottom-up removal of directories. This is implemented
+   * by having a sequence of filenames sorted by decreasing length.  A filename
+   * of greater length must be a subdirectory, or unrelated.
+   *
+   * If a directory isn't empty, that's fine - a package may own
+   * the directory, but it's valid for other packages to put content under it.
+   * Particularly while we're doing a replacement.
```

Comment:
I was going to suggest only doing this if we're doing a replacement, but actually looking at and testing librpm here, it also just leaves the directories there if they're not empty during pure removals. I guess in practice, any packages that drop files in another package's dirs should have a `Requires` on it and so would be removed first anyway.